akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -253,13 +253,14 @@ private[projection] class SlickProjectionImpl[Offset, Envelope, P <: JdbcProfile
 
     // -------------------------------------------------------
     // finally build the source with all parts wired
-    val lastKnownOffset: Future[Option[Offset]] = offsetStore.readOffset(projectionId)
-
-    val futSource = lastKnownOffset.map { offsetOpt =>
-      logger.debug("Starting projection [{}] from offset [{}]", projectionId, offsetOpt)
-      sourceProvider.source(offsetOpt)
+    val readOffsets = () => {
+      val offsetsF = offsetStore.readOffset(projectionId)
+      offsetsF.foreach { offset => logger.debug("Starting projection [{}] from offset [{}]", projectionId, offset) }
+      offsetsF
     }
 
+    val futSource = sourceProvider.source(readOffsets)
+
```

Comment:
Related thought, but for exactlyOnce. The old projection instance is processing offset 17, but before it has been committed to the database there is a hand-over of that topic-partition to another projection instance, reading offset 16, and processing 17. The old instance saves has successfully processed 17 and saves offset 17. The new instance will also process 17. It's not exactly-once.

Any way we can control the hand-over for such scenario?

Reply:
> Related thought, but for exactlyOnce. The old projection instance is processing offset 17, but before it has been committed to the database there is a hand-over of that topic-partition to another projection instance, reading offset 16, and processing 17. The old instance saves has successfully processed 17 and saves offset 17. The new instance will also process 17. It's not exactly-once.
> 
> Any way we can control the hand-over for such scenario?

That's a good point.  If we were using Alpakka Kafka transactions then we could rely on logic within Kafka transactions and extra bookkeeping done in Alpakka Kafka to ensure inflight messages from a revoked partition aren't committed, or all committed before the partition is revoked.

The `Source` we're using right now doesn't have any of these guarantees and is really only suitable to guarantee at-least-once.  We could implement an Alpakka Kafka `PartitionAssignmentHandler` that lets us perform blocking operations during a rebalance.  Such as waiting for the partitioned `Source` for a revoked partition complete, before letting the rebalance process continue (or cancelling it pre-emptively).  This will require some not-so-fun concurrency locking code to implement though.  But since we control the streams implementation it might be an option in projections.

These weird partition rebalancing problems come up often with Alpakka Kafka and result in a lot of the project's complexity.  I'll create an issue to consider solutions for this case.

Reply:
https://github.com/akka/akka-projection/issues/98